def.bzl

```diff
@@ -69,7 +69,58 @@ def _rlocation_path(ctx, file):
     else:
         return ctx.workspace_name + "/" + file.short_path
 
-def _gazelle_runner_impl(ctx):
+def _gazelle_runner_attr_factory(test_runner = False):
+    attrs = {
+        "gazelle": attr.label(
+            allow_single_file = True,
+            default = "//cmd/gazelle",
+            executable = True,
+            cfg = "exec",
+        ),
+        "command": attr.string(
+            values = [
+                "fix",
+                "update",
+                "update-repos",
+            ],
+            default = "update",
+        ),
+        "mode": attr.string(
+            values = ["", "print", "fix", "diff"],
+            default = "",
+        ),
+        "external": attr.string(
+            values = ["", "external", "static", "vendored"],
+            default = "",
+        ),
+        "build_tags": attr.string_list(),
+        "prefix": attr.string(),
+        "extra_args": attr.string_list(),
+        "data": attr.label_list(allow_files = True),
+        "env": attr.string_dict(),
+        "_repo_config": attr.label(
+            default = "@bazel_gazelle_go_repository_config//:WORKSPACE" if GAZELLE_IS_BAZEL_MODULE else None,
+            allow_single_file = True,
+        ),
+        "_template": attr.label(
+            default = "//internal:gazelle.bash.in",
+            allow_single_file = True,
+        ),
+    }
+    if test_runner:
+        attrs.update({
+            "workspace": attr.label(
+                allow_single_file = True,
+                doc = "Label of the WORKSPACE file, required",
+            ),
+            "mode": attr.string(
+                values = ["diff"],
```

Comment:
This setting only has a single value, could we just remove it with `.pop` instead?

Reply:
This is to keep the working style consistent with gazelle. Although only "diff" is allowed, it is clear.